### ISSUE TITLE
NPTicket signature verification

### ISSUE THREAD

**Issue by Slendy** [LBPUnion/ProjectLighthouse#569]
NPTickets, which are used by the game to authenticate with the game server, contain a cryptographic signature of its contents. This signature can be used to verify that a ticket was indeed generated by the expected source and hasn't been tampered with. Fortunately for us, RPCN is open source which means verifying those tickets is relatively simple. The challenge is verifying tickets generated by PSN because there is no official documentation and the tickets are generated server-side so you can't simply reverse engineer the console firmware.

# What is the signature?
`30 36 02 19 00 CC 99 5F 2D 33 4A 4C 2B AD BB 07 1D 21 DF A1 05 BF 12 DC F1 58 07 E8 41 02 19 00 D3 A3 AA C3 4A 20 23 D9 04 2A 7B F2 E3 5A DE 2F E5 B5 41 BA 48 13 37 41`
The signature is located in the footer section of the ticket and is two DER-encoded integers that are roughly 192 bits in length. I believe that this is an ECDSA signature and the two integers correspond to r and s respectively. Theoretically, all we would need to verify this signature is the public key of the signer which in this case is PSN. Due to the nature of ECC, it is possible to recover the public key from just the signature and message but this requires knowledge of the elliptic curve properties, the hashing algorithm, and the exact message bytes.

# What hasn't worked
I have already tried using common 192 bits curves like `secp192k1` and `secp192r1` with common hash algorithms like `SHA1`, `SHA224`, and `SHA256`. Using a sample size of a few hundred tickets I was unable to produce any results using [a small C# program I wrote](https://github.com/Slendy/PubKeyFinder). As a proof of concept, this program is able to recover the public key of signatures from RPCN which means that PSN tickets are either using an uncommon or nonstandard curve type, are using an uncommon hashing algorithm, or the hashed message is more than just the body section of the ticket verbatim. 

# Why does this matter?
If we could cryptographically verify NPTickets then it would mean that Project Lighthouse would no longer have to rely on an IP-based authentication scheme which can be a source of frustration for users with dynamic IP addresses. It could also light the way for a system that is more akin to the original servers where you don't even have to register to connect to the game server. Looking beyond Project Lighthouse, advancements in ticket verification could benefit other private server projects and make them more secure.

# What can I do?
If you read this far and aren't a programmer or cryptography guru and you know anyone who is familiar with the inner workings of PSN, NPTickets or is knowledgeable of elliptic curve cryptography and ECDSA, then link them to this issue and ask if they can help. This is a somewhat complex issue and we could use all the help we can get.

**Comment by TorutheRedFox** [comment]
it'd be great if we could work with RPCN devs on this one to verify the tickets against a specific RPCN server so that you can't just host your own instance and get a valid ticket for a particular username

**Comment by Slendy** [comment]
> it'd be great if we could work with RPCN devs on this one to verify the tickets against a specific RPCN server so that you can't just host your own instance and get a valid ticket for a particular username

This is already done and working in https://github.com/Slendy/ProjectLighthouse/tree/ticket-auth. Since the tickets are signed using an official RPCN key , you can’t use any other server otherwise the signature won’t verify.

**Comment by Cryptofowl** [comment]
I found this repo https://github.com/jmesmon/ps3keys which holds known PS3 key files and curve types. It's intended for use with https://github.com/daryl317/fail0verflow-PS3-tools. Going to spend some time taking a deeper look into them and see if this clears up any more about how they handle signatures, but from what I'm seeing so far it does appear to be ECDSA using SHA-1.

**Comment by Cryptofowl** [comment]
My one outstanding question for this solution is how do we actually validate public keys? Are we building our own database of public keys just assuming the first one we recover belongs to the appropriate user? I could be misunderstanding but since PSN's infrastructure is what actually holds the keys then we have no way of actually checking that the recovered key matches the known public key without duplicating that infrastructure somehow.

**Comment by Slendy** [comment]
What I’ve been doing is I collected a large sample of NPTickets then ran the recovery algorithm on all of them and collected the points it found. Then, if it actually found the correct public key there should be duplicate points with a count close or equal to the sample size. I don’t think that it would be based on a console key because the game server has to be able to verify it and it seems too complex to add a network trip to a Sony api to validate a ticket. The point of embedding a signature like this is so that the client (in this case, the game server) can verify on its own. What I’ve heard from other people is that different games use different keys to sign the ticket.

**Comment by Cryptofowl** [comment]
That makes sense, that will definitely reduce the complexity of reverse engineering the public key for testing. Maybe I'm misunderstanding how the signature is being used here, I'm thinking of it as though each key corresponds to a user. Are you saying that there's singular game keys being used to sign tickets by PSN as a way to prove that the rest of the ticket data is legitimate? So we only need that key to validate the ticket, and can use the rest of the ticket data to authenticate users?

**Comment by Slendy** [comment]
Yeah, the ticket footer also contains 4 bytes of binary data that I’m 99% sure are the same for all the LBP games which would make sense because they all use the same server. According to this [ticket format documentation](https://github.com/hallofmeat/Skateboard3Server/blob/master/docs/PS3Ticket.md) done by the skate 3 custom server team, the 4 bytes are a cipher id that can be used by the game server to know which key to verify with.

**Comment by RipleyTom** [comment]
Signature is verified by hashing everything from beginning of ticket to size of first blob + 12 with SHA1(160 bit). This is then used to verify an ECDSA key which starts just after what is hashed(so after 30 02 00 44). From what I can tell what was assumed to be part of the header is actually part of the signature.

**Comment by Slendy** [comment]
I was able to successfully recover the public key for the LBP games. The curve type is `secp192r1`, the hash method is `SHA1` and the corresponding public key is
x: `39c62d061d4ee35c5f3f7531de0af3cf918346526edac727`
y: `a5d578b55113e612bf1878d4cc939d61a41318403b5bdf86`
The boundary of the message that gets hashed is all the way from the beginning of the ticket to right before the start of the signature. Even the section header for the signature (the 00 08 00 38) is included. One other issue I ran into is that since these tickets are fixed length, sometimes the signature can be less than 56 bytes in which case there will be an extra one or two 0s at the end which needs to be trimmed. Thanks to @RipleyTom for the insight about the ticket message bounds and to everyone else who contributed. I'll be keeping this issue open until this feature is merged into main.
